app/services/immediate_email_generation_service/batch.rb

```diff
@@ -0,0 +1,89 @@
+class ImmediateEmailGenerationService
+  class Batch
+    def initialize(content, subscription_ids_by_subscriber)
+      @content = content
+      @subscription_ids_by_subscriber = subscription_ids_by_subscriber
+    end
+
+    def generate_emails
+      email_ids = []
+      return email_ids unless email_parameters.any?
+
+      ActiveRecord::Base.transaction do
+        email_ids = ContentChangeEmailBuilder.call(email_parameters).ids if content_change
+        email_ids = MessageEmailBuilder.call(email_parameters).ids if message
+
+        records = subscription_content_records(email_ids)
+        SubscriptionContent.populate_for_content(content, records)
+      end
+
+      email_ids
+    end
+
+  private
+
+    attr_reader :subscription_ids_by_subscriber, :content
+
+    def email_parameters
+      @email_parameters ||= begin
+        subscription_ids_by_subscriber
+          .each_with_object([]) do |(subscriber_id, subscription_ids), memo|
```

Comment:
I've moved this code around a bit so this same `each_with_object([])` doesn't apply. However, in this situation a flat_map wouldn't be appropriate as we're selectively building a way so it's a `reduce` situation. If we were using Ruby 2.7 this would be a good place to use `filter_map`.

Reply:
Yes, that would have been a useful method here. The new version is clearer 👍